plugin/sublime-text/c_api_subl3.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (C) 2020  Matthew "strager" Glazar
+# See end of file for extended copyright information.
+
+"""Creates communication between plugin and C++ code."""
+
+import ctypes
+import os
+import zipfile
+
+import sublime
+
+
+class SeverityEnumeration:
+    ERROR = 1
+    WARNING = 2
+
+
+class DiagnosticStructure(ctypes.Structure):
+    """Diagnostic layer used to communicate with the C++ code."""
+
+    # struct qljs_sublime_text_3_diagnostic {
+    #   const char* message;
+    #   const char* code;
+    #   qljs_severity severity;
+    #   int begin_offset;
+    #   int end_offset;
+    # };
+    _fields_ = [
+        ("message", ctypes.c_char_p),
+        ("code", ctypes.c_char_p),
+        ("severity", ctypes.c_int),
+        ("begin_offset", ctypes.c_int),
+        ("end_offset", ctypes.c_int),
+    ]
+
+
+DiagnosticStructurePointer = ctypes.POINTER(DiagnosticStructure)
+
+
+class ParserStructure(ctypes.Structure):
+    """Parser layer used to communicate with the C++ code."""
+
+    _fields_ = []
+
+
+ParserStructurePointer = ctypes.POINTER(ParserStructure)
+
+
+def get_script_path_directory():
+    return os.path.dirname(os.path.realpath(__file__))
+
+
+def get_parent_directory(path, level=1):
+    return os.path.normpath(os.path.join(path, *([".."] * level)))
+
+
+def load_library():
+    # If the plugin is inside a package you will need to unzip the
+    # shared library from inside the package before continuing.
+    #
+    # From:
+    # |-- quick-lint-js-v020-linux_x86_64.sublime-package
+    #
+    # To:
+    # |-- quick-lint-js-v020-linux_x86_64.sublime-package
+    # |-- quick-lint-js
+    #     |-- libquick-lint-js-lib.so
+    #
+    script_path_directory = get_script_path_directory()
+    if script_path_directory.endswith(".sublime-package"):
+        with zipfile.ZipFile(script_path_directory, mode="r") as package_file:
+            lib_path_directory = get_parent_directory(script_path_directory)
+            lib_path_directory += "/quick-lint-js"
+            package_file.extract(
+                "libquick-lint-js-lib.so",
+                path=lib_path_directory,
+            )
```

Comment:
When trying to answer your questions, I found something interesting. But first I need to explain a few things.

**Sublime Text** has its package manager called [Package Control]. It is responsible for installing, updating, and uninstalling packages. Your package format is a **zip file** with the extension `.sublime-package`.

The installation process can be done through **Package Control** or not. To install through it, it is necessary that the package maintainer has added the package/plugin to the list in the [Package Control Default Channel] and that the user has installed the **Package Control**. Installing **Package Control** is very easy as the installation option is available inside **Sublime Text**. If not installed through it, so the installation process will be **manual**. To Install a **directory package** manually, you need to copy it to the path <code><i><data_path></i>/Packages/</code>, and a **zip package** manually (the same format used by **Package Control**), you need to copy it to <code><i><data_path></i>/Installed Packages/</code>. The **zip package** may not work due itself may contain shared libraries and executables.

Now comes what I found interesting, the [Package Control Documentation] says that if you add a file called `.no-sublime-package` to the package root, the **Package Control** will extract the package and install it as a **directory**. When I ask it to uninstall this same package, it will remove the **directory**.

So I'll make some code modifications to achieve this behavior. But it is relevant to make it clear that the automatic extraction is done by **Package Control** only works if the user installs the package through **Package Control**. If the user installs the package through **CMake**, then **CMake** will be responsible for the correct package installation, that would be in this plugin that we're developing, a simple **directory's manual installation** instead of a **zip file's manual installation**.

I **tested** the **Package Control's features** mentioned above, and they worked correctly.

[Package Control](https://packagecontrol.io/)
[Package Control Default Channel](https://github.com/wbond/package_control_channel)
[Package Control Documentation](https://packagecontrol.io/docs)
[Package Control Customizing Packages](https://packagecontrol.io/docs/customizing_packages)
[Package Control Usage](https://packagecontrol.io/docs/usage)
[Package Control Submitting a Package](https://packagecontrol.io/docs/submitting_a_package)
[Sublime Text Packages](https://www.sublimetext.com/docs/packages.html)

[Package Control]: https://packagecontrol.io/
[Package Control Default Channel]: https://github.com/wbond/package_control_channel
[Package Control Documentation]: https://packagecontrol.io/docs/customizing_packages

Reply:
Code modifications: 4078572eb04c3972b39b09bd40d447319dcab248

Reply:
Nice. `.no-sublime-package` looks perfect for us.